### PR TITLE
Zonotope order reduction (Girard's algorithm)

### LINKED ISSUE #195: Order reduction for Zonotopes

### PR DESCRIPTION
**Pull request by mforets**
Closes #195.

### CODE REVIEW

File: src/Zonotope.jl [hunk 1]
```diff
@@ -319,3 +321,66 @@ function scale(α::Real, Z::Zonotope)
...
+### Output
+
+A new zonotope with less generators, if possible.
+
+### Algorithm
+
+This function implements the algorithm described in A. Girard's
+*Reachability of Uncertain Linear Systems Using Zonotopes*, HSCC. Vol. 5. 2005.
+"""
+function reduce_order(Z::Zonotope, r::Int)
```
Comment (schillic) [inline src/Zonotope.jl]: output type?

File: src/Zonotope.jl [hunk 2]
```diff
@@ -319,3 +321,66 @@ function scale(α::Real, Z::Zonotope)
...
+    c, G = Z.center, Z.generators
+    p, d = ngens(Z), dim(Z)
+    h = zeros(p)
```
Comment (schillic) [inline src/Zonotope.jl]: `zeros(N, p)` (you need to define `N` first).

File: src/Zonotope.jl [hunk 3]
```diff
@@ -319,3 +321,66 @@ function scale(α::Real, Z::Zonotope)
...
+    for i in 1:p
+        h[i] = vecnorm(G[:, i], 1) - vecnorm(G[:, i], Inf)
+    end
+    ind = sortperm(h)
+
+    if r * d >= p
```
Comment (schillic) [inline src/Zonotope.jl]: Pull up as far as possible? No need to compute `h` in that case!

File: src/Zonotope.jl [hunk 4]
```diff
@@ -319,3 +321,66 @@ function scale(α::Real, Z::Zonotope)
...
+        # do not reduce
+        return Z
+    end
+    m = p - floor(Int, d * (r - 1))
+    rg = G[:, ind[1:m]]
+
+    # interval hull computation of reduced generators
+    d = sum(abs.(rg), 2)
```
Comment (schillic) [inline src/Zonotope.jl]: What does this line do? `d` is never used again...

File: src/Zonotope.jl [hunk 5]
```diff
@@ -319,3 +321,66 @@ function scale(α::Real, Z::Zonotope)
...
+    Gbox = diagm(sum(abs.(rg), 2)[:])
+    if m > p
+        Gred = [G[:, ind[m+1]:end] Gbox]
+    else
+        Gred = Gbox
+    end
+    Zred = Zonotope(c, Gred)
+    return Zred
```
Comment (schillic) [inline src/Zonotope.jl]: return immediately?

File: src/Zonotope.jl (hunk 2 above)
Comment (schillic) [reply src/Zonotope.jl]: I checked it: For `Rational` it returns `Float64`, but for `Float32` it returns `Float32`. So let us keep it as is (`Float32` is automatically promoted to `Float64`, I checked it).

File: src/Zonotope.jl [hunk 6]
```diff
@@ -319,3 +321,66 @@ function scale(α::Real, Z::Zonotope)
...
+
+This function implements the algorithm described in A. Girard's
+*Reachability of Uncertain Linear Systems Using Zonotopes*, HSCC. Vol. 5. 2005.
+"""
+function reduce_order(Z::Zonotope, r::Int)
+    c, G = Z.center, Z.generators
+    p, d = ngens(Z), dim(Z)
+    h = zeros(p)
+    for i in 1:p
+        h[i] = vecnorm(G[:, i], 1) - vecnorm(G[:, i], Inf)
```
Comment (schillic) [inline src/Zonotope.jl]: For vectors `norm` and `vecnorm` should be equivalent. Any reason why you used `vecnorm` here?

File: src/Zonotope.jl (hunk 4 above)
Comment (mforets) [reply src/Zonotope.jl]: the sum is used just below, to build Gbox

File: src/Zonotope.jl (hunk 6 above)
Comment (mforets) [reply src/Zonotope.jl]: not particular, no. i tried to use `vecnorm` column-wise in the matrix `G` with a broadcast but it didn't work, thus the for loop.

File: src/Zonotope.jl (hunk 4 above)
Comment (schillic) [reply src/Zonotope.jl]: Yes, but this line can be removed I think (or use `d` below instead).

File: src/Zonotope.jl [hunk 7]
```diff
@@ -319,3 +321,66 @@ function scale(α::Real, Z::Zonotope)
...
+        return Z
+    end
+    m = p - floor(Int, d * (r - 1))
+    rg = G[:, ind[1:m]]
+
+    # interval hull computation of reduced generators
+    d = sum(abs.(rg), 2)
+    Gbox = diagm(sum(abs.(rg), 2)[:])
+    if m > p
+        Gred = [G[:, ind[m+1]:end] Gbox]
```
[the remark this answers is not in the document]
Comment (mforets) [reply src/Zonotope.jl]: `ind[m+1]` points to a number (index), so this chooses from that index to the last one

File: src/Zonotope.jl (hunk 4 above)
Comment (mforets) [reply src/Zonotope.jl]: done

File: src/Zonotope.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (mforets) [reply src/Zonotope.jl]: but `ind` is a vector of length `p` thus the check is nonsense (we need `m+1 <= p`), updated.

[the remark this answers is not in the document]
Comment (schillic) [reply src/Zonotope.jl]: equivalent to `m < p` for integers ;)

[the remark this answers is not in the document]
Comment (mforets) [reply src/Zonotope.jl]: actually, `ind[m+1:end]` because these is the set of non reduced generators, well done.

**Pull request merged**